### PR TITLE
Tensorboard view

### PR DESCRIPTION
**Pull request by shark8me**
Hi,

I've created a test/example that compares the performance of different networks and displays the results in a Tensorflow (Tensorboard) compatible log file. This needs an additional library ([tfevent-sink](https://github.com/shark8me/tfevent-sink)), as well as some changes to the listener (already available in [this PR](https://github.com/thinktopic/cortex/pull/169)).

In short, different metrics are evaluated per epoch and sent to a core.async channel as a Clojure map. A receiver reads messages on the channel, creates a TFRecord (Tensorflow Record) event and writes to a log file. Tensorboard reads this file (or multiple event files) and displays graphs of the same.

Also attached is an example of the output (as generated by the example in the PR), as seen on Tensorboard 
tensorboard

### CODE REVIEW

**Comment by cnuernber** [comment]
This really looks great, I am very impressed.

**Comment by shark8me** [comment]
Thanks!
What would be the right way to integrate it with Cortex? There is a dependency on the tfevent-sink library (in the experiment project, not the core cortex project).

File: experiment/test/cortex/experiment/logistic_test.clj [hunk 1]
```diff
@@ -45,5 +45,5 @@
         input-data [{:data [5000.0 10.0]} {:data [5.0 100000.0]}]
         [[should-def] [shouldnt-def]] (->> (execute/run trained-network input-data)
                                            (map :labels))]
-    (is (> should-def 0.97))
+    (is (> should-def 0.96))
```
Comment (harold) [inline experiment/test/cortex/experiment/logistic_test.clj]: o.0

File: experiment/project.clj [hunk 2]
```diff
@@ -3,6 +3,7 @@
   :dependencies [[org.clojure/clojure "1.8.0"]
                  [thinktopic/cortex "0.9.9-SNAPSHOT"]
                  [thinktopic/think.image "0.4.10"]
+                 [org.shark8me/tfevent-sink "0.1.2-SNAPSHOT"]
```
Comment (harold) [inline experiment/project.clj]: Can you make a non-snapshot version of this library?

We prefer not to depend on 3rd party snapshots.

File: experiment/test/cortex/experiment/tensorboard_test.clj [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          train-fn (partial train-and-eval train-ds test-ds)]
+      (mapv train-fn [linear-with-batch-norm
+                      linear-without-batch-norm relu-with-batch-norm
+                      tanh-with-batch-norm]
+            (mapv #(let [path (str log-path % "/tfevents." % ".out")]
+                     (io/make-parents path)
+                     path)
+                  ["linear"
+                   "linear_nobatchnorm" "relu" "tanh"]))
+      (is (.exists (io/file (str log-path "/linear/tfevents.linear.out"))))))
```
Comment (harold) [inline experiment/test/cortex/experiment/tensorboard_test.clj]: This is certainly interesting, and runs to completion without errors for me.

It's definitely a great example of integrating your library into cortex, but I'm not sure how this fits into our visualization strategy more broadly.

Can you include some instructions (or a pointer to some instructions) on how to go from the contents of `/tmp/tflogs` to the attractive screenshot you posted? Would such instructions belong in the README of the `org.shark8me/tfevent-sink` repo?

**Comment by harold** [comment]
I left a few comments, hopefuly they're helpful.

I'll be out until Jun 8, @cnuernber may have opinions on this, or perhaps we can revisit this then and see where it stands.

File: experiment/test/cortex/experiment/logistic_test.clj (hunk 1 above)
Comment (shark8me) [inline experiment/test/cortex/experiment/logistic_test.clj]: On my machine, it worked fine at 0.97. On the Travis CI build, it failed as the as the score was 0.96x.

File: experiment/project.clj (hunk 2 above)
Comment (shark8me) [inline experiment/project.clj]: Sure, I'm working on it.

File: experiment/test/cortex/experiment/tensorboard_test.clj (hunk 3 above)
Comment (shark8me) [inline experiment/test/cortex/experiment/tensorboard_test.clj]: In my experience using Tensorflow (and Tensorboard), I found that there's a learning curve associated with understanding why a network doesn't train, and the Tensorboard UI certainly helps illuminate some of the internals.
Users new to Cortex (and familiar with Tensorboard) might appreciate the familiarity and therefore have a softer learning curve. 
    
Of course Cortex can have a different visualization strategy, and the suggestion to use Tensorboard was not to replace any other visualization, but merely to offer one set of users (those familiar with Tensorboard) a familiar interface. 

I'll add instructions to tfevent-sink on how Tensorboard can launch and view the logs.

File: experiment/test/cortex/experiment/logistic_test.clj (hunk 1 above)
Comment (charlesg3) [inline experiment/test/cortex/experiment/logistic_test.clj]: Fails (barely) on master as well.

```
lein test :only cortex.experiment.logistic-test/logistic-test

FAIL in (logistic-test) (logistic_test.clj:48)

expected: (> should-def 0.97)

  actual: (not (> 0.969944953918457 0.97))

Ran 1 tests containing 2 assertions.

1 failures, 0 errors.

Tests failed.

```

**Comment by shark8me** [comment]
Updated:

- the dependency (tfevent-sink) is updated to use a release version
- documentation added (on tfevent-sink) to load the generated log files using Tensorboard.

**Comment by cnuernber** [comment]
@shark8me:  Rock on, nice work :).

**Comment by shark8me** [comment]
Thank you! :)

**Pull request merged**